### ISSUE TITLE
[Docs]: for useSearchParams, functional updates should be explained in more detail

### ISSUE THREAD

**Issue by YakovL** [remix-run/react-router#11449]
### Describe what's incorrect/missing in the documentation

Looking at https://reactrouter.com/en/main/hooks/use-search-params, one may think that to update one param while keeping the others, one can do:

    const [searchParams, setSearchParams] = useSearchParams()

    //...
    const setPageNumber = (pageNumber: number) => setSearchParams(params => ({ ...params, page: String(pageNumber) }))

to update just the `page` portion of the query. However, this doesn't work: it removes all the query params and sets _just_ the `page` portion. (not sure if this is a bug or an intended behavior)

To work around this, one has to do:

    const setPageNumber = (pageNumber: number) => setSearchParams(params => {
      params.set('page', String(pageNumber))
      return params
    })

I think this is not trivial, and hence should be documented (unless it's a bug, I'm using 6.22.3).

**Comment by kiliman** [comment]
`URLSearchParams` is not a regular object, so you can't spread it like you normally would.

Instead, you need to convert the existing params into an object via `Object.fromEntries()`

```ts
const [searchParams, setSearchParams] = useSearchParams()

//...
const setPageNumber = (pageNumber: number) => 
  setSearchParams(params => ({ ...Object.fromEntries(params), page: String(pageNumber) }))
```
https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams

**Comment by depoulo** [comment]
IMHO the bigger undocumented caveat is that any mutation of `searchParams` (the first array entry returned), e.g. `searchParams.delete('page')` has the side effect of actually changing the URL.

**Comment by YakovL** [comment]
Yeah, both are relevant: the `Object.fromEntries` trick would be helpful there, and then an explanation of how to properly delete a param would be also useful. Is

    setSearchParams(params => {
      const queryParams = Object.fromEntries(params)
      delete queryParams['page']
      return queryParams
    })

sufficient to avoid side-effects?

**Comment by ecline123** [comment]
> IMHO the bigger undocumented caveat is that any mutation of `searchParams` (the first array entry returned), e.g. `searchParams.delete('page')` has the side effect of actually changing the URL.

This is a huge omission, and your comment here is what allowed me to solve a bug in my app that was having me tear out my remaining hair.

**Comment by ryanflorence** [comment]
I've updated the docs on dev to include more usage (https://reactrouter.com/dev/api/hooks/useSearchParams).

Though the particular concern over expecting a `...spread` syntax to work is just a nuance of URLSearchParams, not React Router.

We try not to explain JavaScript/DOM/HTML in our docs because then it gets difficult to know when to stop. It's also distracting for people who already know. Our docs are for React Router, not "web development generally".

We used to explain way too much about HTML in our docs and it we got a lot of feedback that it was distracting (and I agree).

Appreciate you opening the issue though, that doc needed some love (and could probably still use more).

---

@depoulo I can't reproduce this

> IMHO the bigger undocumented caveat is that any mutation of searchParams (the first array entry returned), e.g. searchParams.delete('page') has the side effect of actually changing the URL.


```tsx
import "./App.css";
import { useSearchParams } from "react-router";

function App() {
  const [sp, setSp] = useSearchParams();

  return (
    <>
      <p>
        <button
          onClick={() => {
            setSp(sp => {
              sp.set("now", Date.now() + "");
              return sp;
            });
          }}
        >
          Set Search Param
        </button>{" "}
        <button
          onClick={() => {
            sp.delete("count");
          }}
        >
          Remove Search Param 1
        </button>{" "}
      </p>
      <p>{sp.get("now") || "None"}</p>
    </>
  );
}

export default App;
```

https://stackblitz.com/edit/github-z7wmgfsy?file=app%2Froutes%2Fhome.tsx

I'm closing, but if you can reproduce please open a new issue about it. Thanks everybody!

**Comment by depoulo** [comment]
> if you can reproduce please open a new issue about it

💁‍♂ https://github.com/remix-run/react-router/issues/13493
